python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -1002,3 +1002,96 @@ def with_resources(self, resources: Optional[Mapping[str, Any]]) -> "Definitions
+    ):
+        """Run a provided update function on every contained definition that supports it
+        to updated its metadata. Return a new Definitions object containing the updated objects.
+        """
+        updated_jobs = None
+        if self.jobs is not None:
+            updated_jobs = []
+            for job in self.jobs:
+                if isinstance(job, JobDefinition):
+                    updated_jobs.append(job.with_metadata(update(job.metadata)))
+                elif isinstance(job, UnresolvedAssetJobDefinition):
+                    updated_jobs.append(job.with_metadata(update(job.metadata or {})))
+                else:
+                    check.assert_never(job)
+
+        updated_schedules = None
+        if self.schedules is not None:
+            updated_schedules = []
+            for schedule in self.schedules:
+                if isinstance(schedule, ScheduleDefinition):
+                    updated_schedules.append(
+                        schedule.with_attributes(metadata=update(schedule.metadata))
+                    )
+                elif isinstance(schedule, UnresolvedPartitionedAssetScheduleDefinition):
+                    updated_schedules.append(
+                        schedule.with_metadata(update(schedule.metadata or {}))
+                    )
+                else:
+                    check.assert_never(schedule)
+
+        updated_sensors = None
+        if self.sensors is not None:
+            updated_sensors = []
+            for sensor in self.sensors:
+                if isinstance(sensor, SensorDefinition):
+                    updated_sensors.append(sensor.with_attributes(metadata=update(sensor.metadata)))
+                else:
+                    check.assert_never(sensor)
+
+        updated_assets = None
+        if self.assets is not None:
+            updated_assets = []
+            for asset in self.assets:
+                if isinstance(asset, AssetsDefinition):
+                    updated_assets.append(
+                        asset.with_attributes(
+                            metadata_by_key={
+                                **{
+                                    key: update(metadata)
+                                    for key, metadata in asset.metadata_by_key.items()
+                                },
+                                **{c.key: update(c.metadata) for c in asset.check_specs},
+                            }
+                        )
+                    )
+                elif isinstance(asset, AssetSpec):
+                    updated_assets.append(asset.replace_attributes(metadata=update(asset.metadata)))
+                elif isinstance(asset, (SourceAsset, CacheableAssetsDefinition)):
+                    # these types are deprecated and do not support metadata updates, ignore
+                    updated_assets.append(asset)
+                else:
+                    check.assert_never(asset)
```

Comment:
breaking this up into individual functions per type might allow for early returns and other things to make the indenting less crazy